### DISCUSSION TITLE
RFC #3 - 2.0 Roadmap

### DISCUSSION THREAD

**Discussion by RaddishIoW** [HaylLtd/cordova-plugin-background-geolocation#9] [General]
## Proposal
A roadmap for the next major release (2.0) should be created, to organise which features should be taken forward.

### Discussion Points
* Which features should be dropped, if any?
* Which new features should be proposed.

**Comment by HarelM** [comment]
As this would be a major release I suggest to remove http related code to reduce complexity and maintenance - this would of course introduce breaking changes but since this would be a major release it would be the right time to do it.

**Reply by RaddishIoW** [reply]
Generally, I'm in favour of this, as I think it makes more sense for the app developer using the plugin to worry about the details of what to do with the location data the plugin provides - post it, save it, display it, whatever. The plugin shouldn't be worrying about making http posts...

However - I've been experiencing issues on Android with this plugin since I started using it: I am simply using the event handler to listen for location updates, but I don't believe that I'm getting updates when my app is in the background on Android. I'm yet to check the code, or test this hypothesis, but I think that the background/foreground service on Android only posts location updates to the url if given in the config (i.e. using the http posting provided by the plugin). I don't think it's calling my event handler.
I need to investigate this before we remove the http posting...

**Reply by HarelM** [reply]
I can confirm that's what happening unfortunately.
The webview event handler is begin stopped (I think) and so the js code is not responding to the events.
What I do in order to resolve this is listening to background foreground changes in order to get all valid location and then delete them.
See here:
https://github.com/IsraelHikingMap/Site/blob/27f9e1ae6faad4014326fe8d5ae8cfecbe48983d/IsraelHiking.Web/src/application/services/geo-location.service.ts#L133

My app is a navigation app which works in "offline first" approach so sending location to the server is not even a remotely valid possibility for my app :-)

**Reply by RaddishIoW** [reply]
Hmmm... OK, so I think we might need to leave HTTP posting in, at least for the time being!
I've seen at least one other plugin - this time for Capacitor - which purports to run in the background and does call JS app code when the location changes in the background on Android, so it might be that we can work around this. I'll investigate...

**Reply by HarelM** [reply]
Here's a thread I started in Cordova Slack:
https://cordova.slack.com/archives/C068CHRJ5/p1619934355300300
The person who replied is a core contributor to the Cordova project and plugins and he asks the exact question you did as he is aware of the fact the the webview event handler is being stopped (which is what I thought and now it's confirmed).
In any case, I don't have an objection to keep the http part if you are using it :-)

**Reply by RaddishIoW** [reply]
Couldn't read that thread - I'm guessing I need an invite for that Slack channel to see it?
Anyway, I have seen some odd things while testing the android background stuff in my app... I used to have an event handler for the 'location' event which then posted the location to my server and received the name of the location (e.g. nearest town). This would fire every 5 minutes or so on Android, even when in the background. Sometimes it would fire every 20 seconds, sometimes not for 5 minutes.
I've changed this now so that I'm using the http posting in the plugin to post to my server, and then in the event handler in my main JS when I receive a location update I `GET` from the server for the location name - thinking that the event handler would be called when my app is in the foreground and needed to display the name to the user/driver...
Since the change, I've seen that the event handler *sometimes* gets called even when the app is in the background... i.e. I see the `GET` requests on my server. Doesn't always fire though, but doesn't always fire when in the foreground either!
So sometimes the event handler on the webview is being called, even when in the background, as it was when I was just using the event handler to post location updates...

**Reply by HarelM** [reply]
Well, take into consideration that when you're not moving the location event shouldn't fire. But yes, all-in-all the behavior of this plugin is a bit hard to predict I guess... Feel free to investigate further... I'll be happy to know what you find out...

**Reply by HarelM** [reply]
Just to clarify what I already know:
When the webview is not in front it doesn't mean it is in the background - and thus in this state it still receives events and handle them in the js code. At some point in time the os moves the app to the background and the webview stops responding to events, I think.
It might explain what you are saying I believe.
The term background is confusing as hell...

**Comment by HarelM** [comment]
Proposition: Promises instead of callbacks:
I tried to add a method as part of a feature I needed to add and was reminded that all the methods in this plugin are using callbacks.
I'm using an ionic native wrapper that translates all the callback to promises which is really nice.
If there's a need to go through ionic native team every time we change anything in this plugin it will be a nightmare.
So I suggest to change the javascript interface to uses promises instead of callbacks.
The reasoning for not doing it in Cordova in general is here I think:
https://github.com/apache/cordova-plugin-file/issues/344
Which is no longer relevant I believe...
This is a breaking change so it can't be done for version 1.x...
Let me know what you think...

**Reply by HarelM** [reply]
@RaddishIoW your thoughts?
Maybe we can do like they did in mongodb: when no callbacks are supplied the method will return a promise, this might allow to integrate it in a minor release and not break anything at this moment.

**Reply by RaddishIoW** [reply]
I don't have any objections to it, if it's something that you think would improve the API.
Your suggestion that if no callback functions are supplied then return a promise might work nicely, as long as it doesn't mean too much extra work? Then putting it in a minor version would work. If it would mean too much overhead or work to maintain, then I'm OK with making it a major change.

**Reply by HarelM** [reply]
I think callbacks are annoying in terms of writing code with them. The ionic-native wrappers are a great help in that area, but I'm unable to get the ionic-native maintainers to respond to my calls so the best approach I can think of is simply remove the need for the ionic-native wrapper by incorporating the promise in the js library here. If this will be the case, I basically won't need the ionic wrapper as most of what it currently adds is this promise functionality.
I'll work on it and test it and let you know. I'll add the info in the #16 issue that talks about ionic.

**Comment by wimZ** [comment]
I am successfully using the mauron version at this moment, and have the following build issues. 
1. compile warning, (which I see is fixed in this versions dependencies.gradle)
2. when I remove a plugin, cordova complains that geolocation cannot find a valid package.json, buit still builds correctly.  
So I would prefer to run on a maintained version of the plugin, and this looks a promising project. 

In my app the plugin performs the http post without any problems (or at least I am not aware of it ;) ).  So from that perspective ,at this moment, I do not have a need to remove httppost and would prefer to keep it the way it is. 

I will test the plugin in soon, but would like to know whether you intend to remove http post and if so, which issues would be avoided by doing that. 


Thanks.

**Reply by RaddishIoW** [reply]
Hi wimZ, thanks for your interest!

I'm currently using http posting from this plug-in in my app too, so I think the current plan is to maintain this functionality.

The only reason to remove it would be to reduce the complexity of the plug-in, as far as I'm aware at the moment.
I believe there are some potential issues with the Android implementation, as it uses a legacy library, so we would probably look at upgrading that.

Let us know how you get on with testing this plug-in, and if you have any issues please do get in touch and we'll do our best to help.

Thanks!

**Reply by wimZ** [reply]
I installed the published version ((v1.0.0-rc.5 2021-05-03) just now, and it seems to work correctly without any configuration changes  using httpposts.   Will put it in an alpha test soon to see how it holds up.

**Reply by wimZ** [reply]
It looks good.  Http posts occur with app in background including when phone is locked. 

:)

**Reply by HarelM** [reply]
I suggest to use version 1.0 and not the release candidate...
If Http is needed for your user case there's no reason to remove it.

**Comment by wimZ** [comment]
I honestly have no opinion on this.

I have developed a nice mobile cordova based business app, but do not (yet) use half of javascript features and capabilities.

I'll just hang in there and see what comes.


Kind regards,
wim zandbergen
+31 6 386 84 766

Van: Harel M ***@***.***>
Verzonden: Tuesday, June 22, 2021 10:22
Aan: HaylLtd/cordova-background-geolocation-plugin ***@***.***>
CC: wimZ ***@***.***>; Comment ***@***.***>
Onderwerp: Re: [HaylLtd/cordova-background-geolocation-plugin] RFC #3 - 2.0 Roadmap (#9)


I think callbacks are annoying in terms of writing code with them. The ionic-native wrappers are a great help in that area, but I'm unable to get the ionic-native maintainers to respond to my calls so the best approach I can think of is simply remove the need for the ionic-native wrapper by incorporating the promise in the js library here. If this will be the case, I basically won't need the ionic wrapper as most of what it currently adds is this promise functionality.
I'll work on it and test it and let you know. I'll add the info in the #16<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHaylLtd%2Fcordova-background-geolocation-plugin%2Fissues%2F16&data=04%7C01%7C%7Ceb432a0aee894c202e8008d93556d3ca%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637599469291633908%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=OTS5Z0SJDGBodVnCgwLjRu4hxbd3B2UXPfl66kQ1lro%3D&reserved=0> issue that talks about ionic.

-
